### ISSUE TITLE
design issues in RTL layout

### ISSUE THREAD

**Issue by ShlomoCode** [iina/iina#4776]
After I translated iina into Hebrew (with the kind and patient help of @uiryuu at https://github.com/iina/iina/issues/4742), I note here some display problems that appear only in the Hebrew language which is an RTL (right to left) layout and not in English which is LRT:

- ~~The main video controls (play/pouse) should be on the left ("end of content"), and the others on the right ("start of content")~~ Mistake. see below

- When "Show time and battery information in full screen" is enabled in UI settings, the battery icon is displayed garbled:
![CleanShot 2024-01-16 at 05 40 52@2x](https://github.com/iina/iina/assets/78599753/beebf792-77d8-4c3e-aaad-9d64b87ad59c)

- When there is a scroll bar in the settings, in the RTL layout it should be on the right side of the scrollable content (where I added a yellow line in the picture) and not on the left:
![CleanShot 2024-01-16 at 05 43 47@2x](https://github.com/iina/iina/assets/78599753/de91e2e5-4fc5-46ef-b696-1dd80b3c829e)
this only happens when the system language is LTR, and a RTL language is set for the application in the system settings. when the system language is RTL it is normal, because the buttons are located on the right side.
but this such a mix is quite a common case.

- The buttons that open a color picker look a little damaged in shape:
![CleanShot 2024-01-16 at 05 46 33@2x](https://github.com/iina/iina/assets/78599753/e3f4b3fd-ba7e-4dee-94ad-4c69b3244d5f)

- When adding an audio filter, the original sentence is 2 words "default (ffmage)", in Hebrew the only translation is 3 words, but it should not be cut like in the picture but the column will expand as needed to contain all the words:
![CleanShot 2024-01-16 at 06 23 49@2x](https://github.com/iina/iina/assets/78599753/5dc90779-f246-41dd-a01c-3aa5ae9ea7fc)

- In "Audio delay" in audio panel the number box is too big, and in Hebrew "שנ'" (sec) is cut off
Even if I only translate to the first letter (ש), it is partially cut off
![CleanShot 2024-01-20 at 20 02 47@2x](https://github.com/iina/iina/assets/78599753/d6279790-d495-46c5-b7b3-aa12cb3b3991)

**Comment by uiryuu** [comment]
> The main video controls (play/pouse) should be on the left ("end of content"), and the others on the right ("start of content"):

@ShlomoCode  I'm kinda confused by this. I want to ask some simple questions about some of the control layout/behaviors in the RTL languages.
1. "Play next media" should be a left arrow or a right arrow?
2. The "Play" button should be a left arrow or a right arrow?
3. As for the playback slider (progress bar), when the media is normally played, are you expecting the slider is moving from left to right or right to left?
4. For the volume control, do you expect the left represents higher volume or right represents the higher volume.
5. Why do you think the play back control buttons should be on the left in RTL language?
6. For the sidebar (quicksettings & playlist), do you want them to popup from left instead of from right?

If you can provide answers of these questions, it can help the devs to make better choice to optimize the UI for RTL languages.

**Comment by ShlomoCode** [comment]
@uiryuu 
> 
> 1. "Play next media" should be a left arrow or a right arrow?

left
> 2. The "Play" button should be a left arrow or a right arrow?

left
> 3. As for the playback slider (progress bar), when the media is normally played, are you expecting the slider is moving from left to right or right to left?

From right to left
> 4. For the volume control, do you expect the left represents higher volume or right represents the higher volume.

Left - higher
Right - lower
> 5. Why do you think the play back control buttons should be on the left in RTL language?

It was confusion... :|
they should stay on the right side

> 6. For the sidebar (quicksettings & playlist), do you want them to popup from left instead of from right?

Yes (actually this is the case now thanks to the automatic RTL)


----
The guiding principle is that while in LTR everything flows from left to right, in RTL everything flows from right to left.
This means that in English the start of the window is at the left end and everything flows from there, and in Hebrew the "start" of the window is the right end, and from there everything flows.
Just like when you set the MacOS system language to Hebrew, the apple icon moves to the right corner, and the clock to the left corner.

I see I'm a little unclear, so I edited a screenshot of iina with highlights of the desired changes:
before (now):
![CleanShot 2024-01-24 at 00 47 15@2x copy](https://github.com/iina/iina/assets/78599753/fee7eaf0-f029-4b77-888a-ce5fe81b99ae)

After (the changes are marked with yellow text):
![CleanShot 2024-01-24 at 00 47 15@2x copy 2](https://github.com/iina/iina/assets/78599753/0653f6f3-faf1-4ff8-9ebf-49309c706ecd)

**Comment by ShlomoCode** [comment]
@uiryuu How can we move forward?
I invested many hours in the translation 😶

**Comment by uiryuu** [comment]
Sorry, I was really busy these days. You work is well appreciated, and we will merge the translation PR and deal with layout related issue at appropriate time.

**Comment by ShlomoCode** [comment]
> and we will merge the translation PR

So do I have to submit a PR that activates the Hebrew language?

**Comment by low-batt** [comment]
From the Apple documentation for [NSWindow.windowTitlebarLayoutDirection](https://developer.apple.com/documentation/appkit/nswindow/1644535-windowtitlebarlayoutdirection):
> The layout direction of the window title bar includes the standard window buttons (close, minimize, maximize) and the title for the window. In general, this returns [NSUserInterfaceLayoutDirection.rightToLeft](https://developer.apple.com/documentation/appkit/nsuserinterfacelayoutdirection/righttoleft) if the primary system language is right to left. The layout direction may be right to left even in applications that don’t have a right to left language localization.

I'm not happy when documentation contains the words "may be" without an explanation. It sounds like macOS positions the buttons based on the system configured language and not the application configured language. To me that is a macOS defect. How can you provide the ability to set a language on an application basis and only change half of the application behavior?

As `windowTitlebarLayoutDirection` is a get only property I'm thinking it is not possible for an application to control where the window buttons appear.

I'm a junior IINA developer. Maybe one of the other IINA developers knows a way for an application to change this.

**Comment by ShlomoCode** [comment]
In fact the really big problem is the inversion when dragging:

https://github.com/iina/iina/assets/78599753/202d3c1e-760f-4349-80a5-d2d0e9d172d1

**Comment by low-batt** [comment]
The slider problem reproduced for me. I agree this is a critical problem. Investigating…

**Comment by low-batt** [comment]
Both Quicktime and Safari do not reverse the playback position slider when the language is Hebrew. That is the behavior you are expecting from IINA, yes? If that is the correct behavior the IINA problem is trivial to fix.

Quicktime does reverse the volume slider along with the speaker icon. IINA is failing to reverse that icon. That is another problem that needs to be corrected.

Quicktime also does not change the position of the volume slider or the buttons whereas IINA swaps the positions of the volume slider and buttons. What do you view as the correct behavior?

**Comment by ShlomoCode** [comment]
> Both Quicktime and Safari do not reverse the playback position slider when the language is Hebrew. That is the behavior you are expecting from IINA, yes? If that is the correct behavior the IINA problem is trivial to fix.

It is true that usually video players (including Quicktime) do not invert the playback slider, so such behavior (not inverting, like Quicktime) would be accepted.

But if Quicktime was built for/by a Hebrew speaking person (RTL), he would absolutely build the software with a slider that moves from right to left.
  So if IINA wants to be completely adapted to RTL habits, you should do everything from right to left, as I demonstrated [above](https://github.com/iina/iina/issues/4776#issuecomment-1907066214:~:text=After%20(the%20changes%20are%20marked%20with%20yellow%20text)%3A). (but it's an additional effort that I will accept with understanding if you don't want to invest)

**Comment by low-batt** [comment]
I already looked a little bit into implementing a reversed playback position slider. I believe I can change IINA to properly reverse the control. I'm expecting the mini player and Touch Bar will also need to be updated.

I will be entering a separate issue for this aspect of the RTL problems for tracking purposes. Stay tuned.

**Comment by low-batt** [comment]
I've entered issue #4878 specifically for the OSC related RTL problems.

**Comment by ShlomoCode** [comment]
See https://github.com/iina/iina/issues/4884 for the OSD related RTL issues.

**Comment by low-batt** [comment]
I'm working on these problems now.

**Comment by low-batt** [comment]
On this:
> In "Audio delay" in audio panel the number box is too big, and in Hebrew "שנ'" (sec) is cut off
Even if I only translate to the first letter (ש), it is partially cut off

I'm pretty sure in Hebrew that label should be `s`. It is not a letter or an abbreviation, it is an International System of Units (SI) symbol. From the US Metric Association [FAQ](https://usma.org/faq):
> SI unit symbols are the same in all languages, as you can see on this nutrition label from Japanese snack cakes.

This label should not be presented to translators. I did not find an official Apple way to do that. I found people putting a special string in the `Localizer hint` and editing entries with that hint out of the exported localizations.

Will need to ask @uiryuu if IINA has some sort of convention for this.

At least that is my take on this particular problem.

**Comment by ShlomoCode** [comment]
> SI unit symbols are the same in all languages, as you can see on this nutrition label from Japanese snack cakes.

I don't know what is happening in Japan... in Israel it is mandatory to mark components in Hebrew, and it is customary to translate the units as well. For example mg/ml are always translated into `מ"ג` and `מ"ל`, and even gram is sometimes translated into an abbreviated spelling `גר'` (if it is a force it is usually customary to write "gram" in the full form - `גרם`)
Anyway, you will never see the English-only unit.
That's why it is necessary to translate here also to `ש` or `שנ'` or `שניה` but certainly not to leave only the English measure which is something that is not accepted here.
Attached are some sample product photos:

<img src="https://github.com/iina/iina/assets/78599753/0e149ba4-322b-4bce-b25d-78464562ce9b" width="30%"></img> <img src="https://github.com/iina/iina/assets/78599753/b4025c29-00f7-4240-b893-eb6763cdfce2" width="30%"></img> <img src="https://github.com/iina/iina/assets/78599753/d8fb4031-4d0f-4166-95ac-6cda13db9b71" width="30%"></img> <img src="https://github.com/iina/iina/assets/78599753/8756a979-7111-41bb-b4d8-8991a68a693f" width="30%"></img> <img src="https://github.com/iina/iina/assets/78599753/499cc659-30f0-4a37-874a-f1fd30603f35" width="30%"></img> <img src="https://github.com/iina/iina/assets/78599753/e4d832bf-d43c-4bfd-915e-7a026fe2460c" width="30%"></img>

**Comment by low-batt** [comment]
Understood. I will plan on looking into that overlapping. Currently struggling with the battery and color well problems.

**Comment by ShlomoCode** [comment]
@low-batt The effort you put in is greatly appreciated! Thanks

**Comment by low-batt** [comment]
@ShlomoCode It is all thanks to _you_ that this localization support is being added. It was your PR #4864 that caught my attention and motivated me to help out. I really appreciate your reviews of my proposed changes and the nice way you report my dumb mistakes.

On this one:
> When there is a scroll bar in the settings, in the RTL layout it should be on the right side of the scrollable content (where I added a yellow line in the picture) and not on the left

In the LTR layout the scroll bar is on the right edge of the window. Reversing that would put the scroll bar on the left edge of the window. I'm confused. Why should it be moved to the middle of the window?

I'm currently looking to see if a developer has the time to review and merge PR #4893 and then update Crowdin. That PR adds 3 new strings to be translated. Once Crowdin has been updated I'll be asking you to provide translations for the new strings. Once that is done then I will ask to have the latest translations pulled from Crowdin and merged into `develop`. Next up after that will be reviews and merging of the accumulated PRs. Once that is done it will be easier to spot anything that was missed.

**Comment by ShlomoCode** [comment]
Now I found something - the progress indicator in the "playback history" (cmd + shift + h) is from right to left, when according to what is said in #4878 it should be from left to right.

**Comment by ShlomoCode** [comment]
> On this one:
> 
> > When there is a scroll bar in the settings, in the RTL layout it should be on the right side of the scrollable content (where I added a yellow line in the picture) and not on the left
> 
> In the LTR layout the scroll bar is on the right edge of the window. Reversing that would put the scroll bar on the left edge of the window. I'm confused. Why should it be moved to the middle of the window?

This is actually a form of bypassing what you referred to in https://github.com/iina/iina/issues/4776#issuecomment-2038900007
The Google Chrome browser does this (scroll bar on the right). But I notice that apple apps (Safari for example) don't do that.

System language: English & application language: Hebrew
![CleanShot 2024-04-27 at 23 45 46@2x](https://github.com/iina/iina/assets/78599753/ac63f545-285b-413e-9e6b-492794e40317)


System language: Hebrew & Application language: Hebrew
![CleanShot 2024-04-28 at 01 29 47@2x](https://github.com/iina/iina/assets/78599753/dacd6e27-9f9b-44bf-b633-c18498521802)

**Comment by ShlomoCode** [comment]
Now I noticed - the progress indicator in "playback history" (cmd + shift + h) is from right to left, when according to what was said in #4878 it should be from left to right
![CleanShot 2024-04-28 at 01 44 30@2x](https://github.com/iina/iina/assets/78599753/828244a1-fea6-4cf9-a90b-fa73525f4e8b)

**Comment by ShlomoCode** [comment]
Another issue is the overlapping of the words "default" in the video tab of the quick settings
In Hebrew it cannot be translated into one word, and the 2 words are cut off, while the input is unnecessarily width.

![CleanShot 2024-04-28 at 01 58 21@2x](https://github.com/iina/iina/assets/78599753/b493994f-7ede-4c8b-9922-e643bca6505c)

**Comment by low-batt** [comment]
I will need to take closer look at the window buttons issue. I was already confused as to why the buttons are in the correct place when run under Xcode. That it is working for Safari suggests it is something IINA is doing wrong.

I was expecting we'd be finding additional problems for a while. These all look like problems that need to be corrected. I will keep investigating and proposing fixes.

**Comment by ShlomoCode** [comment]
> That it is working for Safari suggests it is something IINA is doing wrong.

Even in Safari, the buttons are on the left side if the Hebrew language is not set for the entire system. there is simply a "header" area, so there is no overlap.
![CleanShot 2024-04-28 at 02 39 16@2x](https://github.com/iina/iina/assets/78599753/12dea77b-3bd4-47cb-bb06-4d2bf1c8fac6)

**Comment by low-batt** [comment]
Thanks for explaining the window buttons problem. I was confused about the situation with Safari.

I am continuing to work on fixes.

**Comment by ShlomoCode** [comment]
I found a layout problem in the equalizer slider in Quick Settings > Audio tab. The label of one of the sliders is out of place:
![CleanShot 2024-05-05 at 07 06 35@2x](https://github.com/iina/iina/assets/78599753/039498d7-7ac5-4e53-9d64-768320c0d340)
In English it is normal:
![CleanShot 2024-05-05 at 07 06 00@2x](https://github.com/iina/iina/assets/78599753/8705632f-c873-46f7-aff0-766de203e9a7)

**Comment by low-batt** [comment]
I created issue #4923 for the equalizer problem.

**Comment by low-batt** [comment]
The translations on [Crowdin](https://translate.iina.io/) are no longer at 100%.  We have started merging and there is some new text that needs translating. Not sure how much new text will be added for this next release. The one after it is a feature release and definitely will be adding new text.

Have a look at my review of PR #4922.

**Comment by ShlomoCode** [comment]
I have translated the 4 new strings (in fact Hebrew is the only language that is at 100% now lol), and I will try to continue translating in the future when necessary
In any case, I don't think that the lack of a few single strings is a consideration for language activation

**Comment by low-batt** [comment]
Excellent. Yes, does not have to be at 100% for activation. Just thought you would like to be the only language at 100%. 😄

**Comment by ShlomoCode** [comment]
I found a small issue - in the reference to the browser extensions
The arrow should point in the direction of the text, and is hardcoded to the right
![CleanShot 2024-05-23 at 06 21 53@2x](https://github.com/iina/iina/assets/78599753/b11f4265-3b74-4f23-9d9d-2a099762ea09)

By the way, I tried to solve it in Interface Builder, but it seems to be a system icon, and I couldn't find how to set a Mirror for it in RTL. I wonder how to deal with it...
The direction I understood in the meantime is adding a horizontal flip method to the NSImage extension and then using it in PrefUtilsViewController
![CleanShot 2024-05-23 at 07 03 48@2x](https://github.com/iina/iina/assets/78599753/c061bebe-a964-48d6-9b80-0f61d1676a00)

**Comment by seanify-24** [comment]
@ShlomoCode I find this really interesting! Hope you don't mind my inquiry on an old topic. The [Google Material guidelines](https://m3.material.io/foundations/layout/bidirectionality-rtl#88d0cde8-24d1-49c1-88e2-4b65bc5ec3fd) on RTL languages states not to reverse media player UIs, and I noticed YouTube still uses LTR with the media player. I'm wondering if this is something that is debated frequently or locale dependent?

**Comment by ShlomoCode** [comment]
@seanify-24 Yes, see https://github.com/iina/iina/issues/4878

**Comment by low-batt** [comment]
Since the subject of RTL has come up…

What will be IINA 1.5.0 beta 1 is currently under heavy development. One of the goals of this release was to merge major UI work that had been going on in the background as keeping those changes "on the side" was causing double work because changes to the current UI required those changes to then be adapted for the new UI code. That merge has occurred although some portions are not enabled yet. One aspect of this work was to refactor the UI code to not use XIBs. Some portions of the UI are still using XIBs, but significant portions have been changed to use code.

Such major refactoring brings with it the risk of regressions. As significant changes are still being merged the focus has been on finishing coding rather than searching for regressions. I just started testing the `develop` branch looking for regressions and found a few that I need to create issues for. Since XIBs supplied a lot of the RTL related settings, one concern I have is whether there are RTL related regressions. I had planned to bring this up when `develop` was a little more stable. But since the subject has come up, I figured I should mention what is going on.

As others are doing most of this work, I'm unsure as to whether they are ready for feedback on RTL issues. But the changes are in the [Nightly builds](https://iina.io/nightly/) and that makes them fair game for reporting problems.

I have briefly checked Hebrew. One issue is that there is a lot of new text that needs to be translated. I did see at least one thing that looked like a spacing problem.

The part that has not been enabled yet is the new settings window. If you are curious that can be enabled by running this command in terminal:
```text
defaults write com.colliderli.iina enableNewSettings -bool true
```

**Comment by ShlomoCode** [comment]
Here's one thing I noticed right away

<img width="2800" height="829" alt="Image" src="https://github.com/user-attachments/assets/ba5572d9-c902-46e7-a618-da1314f978dc" />

**Comment by low-batt** [comment]
That the "stoplight" buttons are still on the left?

I remember this came up in the past and we concluded this is a macOS problem where AppKit fails to adjust the window properly when the language is set for a specific app and not for the entire system. That can be seen in the screenshots posted in issue #4950.

Was the above screenshot taken when the system was not set to a RTL language?

The new UI does make this problem worse as the sidebar now intrudes into the area normally occupied by the title bar. Notice AppKit correctly left room for the stoplight buttons on the right side of the window. And the icon that indicates the floating on top status is in the correct position to not be hidden by the sidebar.

I think Apple has to fix this.

**Comment by ShlomoCode** [comment]
> I remember this came up in the past and we concluded this is a macOS problem where AppKit fails to adjust the window properly when the language is set for a specific app and not for the entire system. That can be seen in the screenshots posted in issue [#4950](https://github.com/iina/iina/issues/4950).
> 
> Was the above screenshot taken when the system was not set to a RTL language?

Yes exactly

**Comment by low-batt** [comment]
Have a look at issue #6192. I _think_ I have found a viable workaround for this odd AppKit behavior. Will have to see how the other developers feel about overriding AppKit.

That issue and fix is only about the main window problem shown in your comment above. Other IINA windows also may exhibit similar problems. Of course the first step will be to make sure they all work when an RTL language is used at the system level. As the PR only applies the proposed workaround to the main window, it will make that window inconsistent with the others. Possibly reviewers will want the workaround applied to all windows. Stay tuned…
